web_poet/rules.py

```diff
@@ -259,6 +316,50 @@ def search_overrides(self, **kwargs) -> List[ApplyRule]:
         warnings.warn(msg, DeprecationWarning, stacklevel=2)
         return self.search(**kwargs)
 
+    def _rules_for_url(
+        self, url: Union[_Url, str], matchers: Dict[Any, URLMatcher]
+    ) -> Mapping[Type, Type[ItemPage]]:
+        result: Dict[Type, Type[ItemPage]] = {}
+
+        url = str(url)
+
+        for target, matcher in matchers.items():
+            rule_id = matcher.match(url)
+            if rule_id is not None:
+                result[target] = self._rules[rule_id].use
+
+        return result
+
+    def overrides_for(
+        self, url: Union[_Url, str]
+    ) -> Mapping[Type[ItemPage], Type[ItemPage]]:
+        """Finds all of the page objects associated with the given URL and
+        returns a Mapping where the 'key' represents the page object that is
+        **overridden** by the page object in 'value'.
+
+        See example: :ref:`rules-overrides_for-example`.
+        """
+        return self._rules_for_url(url, self._overrides_matchers)
+
+    def page_object_for_item(
+        self, url: Union[_Url, str], item_cls: Type
+    ) -> Optional[Type]:
+        """Return the page object class associated with the given URL that's able
+        to produce the given ``item_cls``.
+
+        See example: :ref:`rules-page_object_for_item-example`.
+        """
+
+        url = str(url)
+
+        matcher = self._item_matchers.get(item_cls)
+        if matcher:
+            rule_id = matcher.match(url)
+            if rule_id is not None:
+                return self._rules[rule_id].use
```

Comment:
It'd be nice to clean up a code a bit: 

* some code is duplicated between page_object_for_item and _rules_for_url
* page_object_for_item and overrides_for are not "symmetric" in implementation: inner code is extracted to a method in overrides_for, but inlined in page_object_for_item.

Reply:
Refactored. https://github.com/scrapinghub/web-poet/pull/112/commits/299f1652521a8e8da956133250cc2501dc552918